MRPT runs on MacOS10.14

Emmm... I have tried to build mrpt on MacOS10.14, but some requirements can not be fullfilled.

Is there any solutions for running MRPT on MacOS? Or does anyone try to fix it? I would like to join this fixed work.

Sorry for my late reply.

There is the error info below.
```
/Users/aliben/project/source_code/mrpt/libs/core/src/aligned_malloc.cpp: In function 'void* mrpt::aligned_malloc(size_t, size_t)':
/Users/aliben/project/source_code/mrpt/libs/core/src/aligned_malloc.cpp:28:11: error: '::aligned_alloc' has not been declared
  return ::aligned_alloc(alignment, size);
           ^~~~~~~~~~~~~
make[2]: *** [libs/core/CMakeFiles/mrpt-core.dir/src/aligned_malloc.cpp.o] Error 1
make[1]: *** [libs/core/CMakeFiles/mrpt-core.dir/all] Error 2
make: *** [all] Error 2
```

I have tried to use the default compiler(Apple provided) and gnu-gcc-8, but the problem is the same.

There is the error info using complier Apple one.

```
Scanning dependencies of target DocumentationFiles
[  0%] Built target DocumentationFiles
Scanning dependencies of target mrpt-random
[  0%] Building CXX object libs/random/CMakeFiles/mrpt-random.dir/src/RandomGenerator.cpp.o
[  0%] Building CXX object libs/random/CMakeFiles/mrpt-random.dir/src/random-precomp.cpp.o
[  0%] Linking CXX shared library ../../lib/libmrpt-random.dylib
[  0%] Built target mrpt-random
Scanning dependencies of target mrpt-core
[  0%] Building CXX object libs/core/CMakeFiles/mrpt-core.dir/src/aligned_malloc.cpp.o
/Users/aliben/project/source_code/mrpt/libs/core/src/aligned_malloc.cpp:28:11: error: 
      no member named 'aligned_alloc' in the global namespace
        return ::aligned_alloc(alignment, size);
               ~~^
1 error generated.
make[2]: *** [libs/core/CMakeFiles/mrpt-core.dir/src/aligned_malloc.cpp.o] Error 1
make[1]: *** [libs/core/CMakeFiles/mrpt-core.dir/all] Error 2
make: *** [all] Error 2
```

Thanks for your comment.

There is a new error info while I added `std` beyond `::aligned_alloc`. It seems like that the std library doesn't have the function `std::aligned_alloc`. It is the same result on gcc(Apple) and gcc-8(gnu). 
```
/Users/aliben/project/source_code/mrpt/libs/core/src/aligned_malloc.cpp:28:14: error: 'aligned_alloc' is not a member of 'std'
  return std::aligned_alloc(alignment, size);
              ^~~~~~~~~~~~~
/Users/aliben/project/source_code/mrpt/libs/core/src/aligned_malloc.cpp:28:14: note: suggested alternative: 'aligned_union'
  return std::aligned_alloc(alignment, size);
              ^~~~~~~~~~~~~
              aligned_union
make[2]: *** [libs/core/CMakeFiles/mrpt-core.dir/src/aligned_malloc.cpp.o] Error 1
make[1]: *** [libs/core/CMakeFiles/mrpt-core.dir/all] Error 2
make: *** [all] Error 2
```

UPDATE:
-------
`#include "core-precomp.h"  // Precompiled headers
#include <mrpt/core/aligned_allocator.h>
#include <cstdlib>  // free, realloc, C++17 aligned_alloc
#include <cstring>  // memset
#ifdef __APPLE__
#include <boost/align/aligned_alloc.hpp>
#endif  // END of __APPLE__

void* mrpt::aligned_calloc(size_t bytes, size_t alignment)
{
	void* ptr = mrpt::aligned_malloc(bytes, alignment);
	if (ptr) ::memset(ptr, 0, bytes);
	return ptr;
}
void* mrpt::aligned_malloc(size_t size, size_t alignment)
{
	// size must be an integral multiple of alignment:
	if ((size % alignment) != 0) size = ((size / alignment) + 1) * alignment;
#ifdef _MSC_VER
	return _aligned_malloc(size, alignment);
#elif __APPLE__
  return boost::alignment::aligned_alloc(alignment, size);
#else
	return ::aligned_alloc(alignment, size);
#endif
}
void mrpt::aligned_free(void* ptr)
{
#ifdef _MSC_VER
	return _aligned_free(ptr);
#else
	return ::free(ptr);
#endif
`
 Error Info

`[  3%] Building CXX object libs/system/CMakeFiles/mrpt-system.dir/src/os.cpp.o
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:692:3: error: 
      unknown type name 'HANDLE'
                HANDLE g_hChildStd_IN_Rd = NULL;
                ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:693:3: error: 
      unknown type name 'HANDLE'
                HANDLE g_hChildStd_IN_Wr = NULL;
                ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:694:3: error: 
      unknown type name 'HANDLE'
                HANDLE g_hChildStd_OUT_Rd = NULL;
                ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:695:3: error: 
      unknown type name 'HANDLE'
                HANDLE g_hChildStd_OUT_Wr = NULL;
                ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:697:3: error: 
      unknown type name 'HANDLE'
                HANDLE g_hInputFile = NULL;
                ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:698:3: error: 
      unknown type name 'SECURITY_ATTRIBUTES'
                SECURITY_ATTRIBUTES saAttr;
                ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:700:27: error: use of
      undeclared identifier 'SECURITY_ATTRIBUTES'
                saAttr.nLength = sizeof(SECURITY_ATTRIBUTES);
                                        ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:701:27: error: use of
      undeclared identifier 'TRUE'
                saAttr.bInheritHandle = TRUE;
                                        ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:705:10: error: use of
      undeclared identifier 'winerror2str'
                        throw winerror2str("StdoutRd CreatePipe");
                              ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:709:49: error: use of
      undeclared identifier 'HANDLE_FLAG_INHERIT'
                if (!SetHandleInformation(g_hChildStd_OUT_Rd, HANDLE_FLAG_I...
                                                              ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:710:10: error: use of
      undeclared identifier 'winerror2str'
                        throw winerror2str("Stdout SetHandleInformation");
                              ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:715:10: error: use of
      undeclared identifier 'winerror2str'
                        throw winerror2str("Stdin CreatePipe");
                              ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:719:48: error: use of
      undeclared identifier 'HANDLE_FLAG_INHERIT'
                if (!SetHandleInformation(g_hChildStd_IN_Wr, HANDLE_FLAG_IN...
                                                             ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:720:10: error: use of
      undeclared identifier 'winerror2str'
                        throw winerror2str("Stdin SetHandleInformation");
                              ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:723:3: error: 
      unknown type name 'PROCESS_INFORMATION'
                PROCESS_INFORMATION piProcInfo;
                ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:724:3: error: 
      unknown type name 'STARTUPINFOA'
                STARTUPINFOA siStartInfo;
                ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:725:3: error: 
      unknown type name 'BOOL'
                BOOL bSuccess = FALSE;
                ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:725:19: error: use of
      undeclared identifier 'FALSE'
                BOOL bSuccess = FALSE;
                                ^
/Users/xxxx/project/source_code/mrpt/libs/system/src/os.cpp:729:34: error: use of
      undeclared identifier 'PROCESS_INFORMATION'
                ZeroMemory(&piProcInfo, sizeof(PROCESS_INFORMATION));
                                               ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [libs/system/CMakeFiles/mrpt-system.dir/src/os.cpp.o] Error 1
make[1]: *** [libs/system/CMakeFiles/mrpt-system.dir/all] Error 2
make: *** [all] Error 2
`

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -60,7 +60,19 @@ jobs:
           command: bash travis/check_style.sh
           environment:
             TASK: lint_all
-
+  macos:
+    macos:
+      xcode: "10.2.0"
+    working_directory: /Users/distiller/project
+    steps:
+      - checkout
+      - run: brew install cmake
+      - run: brew install boost
+      - run: brew install nasm
+      - run: cmake -DBUILD_EXAMPLES=On -DBUILD_TESTING=On -DENABLE_COVERAGE=On -DDISABLE_PYTHON_BINDINGS=On -DEIGEN_USE_EMBEDDED_VERSION=On -H. -Bbuild      
+      - run: make -j4 -C build tests_build_all
+      - run: make -j4 CTEST_OUTPUT_ON_FAILURE=1 test -C build
+      - run: make -j4 -C build
 workflows:
   version: 2
   build:
@@ -70,3 +82,4 @@ workflows:
       - minimal
       - clang
       - clang-format-lint
+      - macos

--- a/cmakemodules/DeclareMRPTLib.cmake
+++ b/cmakemodules/DeclareMRPTLib.cmake
@@ -32,7 +32,12 @@ function(mrpt_lib_target_requires_cpp17 _TARGET)
 	else()
 		if (NOT MSVC)
 			# Modern, clean way to do this:
-			target_compile_features(${_TARGET} PUBLIC cxx_std_17)
+			get_target_property(target_type ${_TARGET} TYPE)
+			if (${target_type} STREQUAL "INTERFACE_LIBRARY")
+				target_compile_features(${_TARGET} INTERFACE cxx_std_17)
+			else()
+				target_compile_features(${_TARGET} PUBLIC cxx_std_17)
+			endif()
 		else()
 			# At present (CMake 3.12 + MSVC 19.15.26732.1) it seems cxx_std_17
 			# does not enable C++17 in MSVC (!).

--- a/cmakemodules/DeclareMRPTLib.cmake
+++ b/cmakemodules/DeclareMRPTLib.cmake
@@ -32,7 +32,7 @@ function(mrpt_lib_target_requires_cpp17 _TARGET)
 	else()
 		if (NOT MSVC)
 			# Modern, clean way to do this:
-			target_compile_features(${_TARGET} INTERFACE cxx_std_17)
+			target_compile_features(${_TARGET} PUBLIC cxx_std_17)
 		else()
 			# At present (CMake 3.12 + MSVC 19.15.26732.1) it seems cxx_std_17
 			# does not enable C++17 in MSVC (!).

--- a/libs/containers/include/mrpt/containers/ts_hash_map.h
+++ b/libs/containers/include/mrpt/containers/ts_hash_map.h
@@ -12,6 +12,7 @@
 #include <mrpt/core/common.h>  // remove MSVC warnings
 #include <array>
 #include <stdexcept>
+#include <string_view>
 
 namespace mrpt::containers
 {

--- a/libs/core/include/mrpt/core/bits_math.h
+++ b/libs/core/include/mrpt/core/bits_math.h
@@ -8,7 +8,7 @@
    +------------------------------------------------------------------------+ */
 #pragma once
 
-#include <algorithm> // max()
+#include <algorithm>  // max()
 #include <cmath>  // floor(),isnan(),...
 #include <stdexcept>
 

--- a/libs/core/include/mrpt/core/bits_math.h
+++ b/libs/core/include/mrpt/core/bits_math.h
@@ -8,6 +8,7 @@
    +------------------------------------------------------------------------+ */
 #pragma once
 
+#include <algorithm> // max()
 #include <cmath>  // floor(),isnan(),...
 #include <stdexcept>
 

--- a/libs/core/src/aligned_malloc.cpp
+++ b/libs/core/src/aligned_malloc.cpp
@@ -24,6 +24,12 @@ void* mrpt::aligned_malloc(size_t size, size_t alignment)
 	if ((size % alignment) != 0) size = ((size / alignment) + 1) * alignment;
 #ifdef _MSC_VER
 	return _aligned_malloc(size, alignment);
+#elif __APPLE__
+	void* p;
+	if (::posix_memalign(&p, alignment, size) != 0) {
+		p = 0;
+	}
+	return p;
 #else
 	return ::aligned_alloc(alignment, size);
 #endif

--- a/libs/core/src/aligned_malloc.cpp
+++ b/libs/core/src/aligned_malloc.cpp
@@ -26,7 +26,8 @@ void* mrpt::aligned_malloc(size_t size, size_t alignment)
 	return _aligned_malloc(size, alignment);
 #elif __APPLE__
 	void* p;
-	if (::posix_memalign(&p, alignment, size) != 0) {
+	if (::posix_memalign(&p, alignment, size) != 0)
+	{
 		p = 0;
 	}
 	return p;

--- a/libs/maps/CMakeLists.txt
+++ b/libs/maps/CMakeLists.txt
@@ -38,7 +38,7 @@ if(BUILD_mrpt-maps)
 	endif()
 
 	if (CMAKE_MRPT_HAS_VTK)
-        include(${VTK_USE_FILE})
+		include(${VTK_USE_FILE})
 		target_link_libraries(maps PRIVATE ${VTK_LIBRARIES})
 	endif()
 
@@ -52,4 +52,5 @@ if(BUILD_mrpt-maps)
 		target_link_libraries(maps PRIVATE ${OCTOMAP_LIBRARIES})
 	endif()
 
+	target_link_libraries(maps PRIVATE ${MRPT_OPENGL_LIBS})
 endif()

--- a/libs/math/src/CMatrix.cpp
+++ b/libs/math/src/CMatrix.cpp
@@ -25,7 +25,7 @@ uint8_t CMatrix::serializeGetVersion() const { return 0; }
 void CMatrix::serializeTo(mrpt::serialization::CArchive& out) const
 {
 	// First, write the number of rows and columns:
-	out << (uint32_t)rows() << (uint32_t)cols();
+	out << static_cast<uint32_t>(rows()) << static_cast<uint32_t>(cols());
 
 	if (rows() > 0 && cols() > 0)
 		for (Index i = 0; i < rows(); i++)
@@ -59,8 +59,8 @@ void CMatrix::serializeFrom(mrpt::serialization::CArchive& in, uint8_t version)
 void CMatrix::serializeTo(mrpt::serialization::CSchemeArchiveBase& out) const
 {
 	SCHEMA_SERIALIZE_DATATYPE_VERSION(1);
-	out["nrows"] = this->rows();
-	out["ncols"] = this->cols();
+	out["nrows"] = static_cast<uint32_t>(this->rows());
+	out["ncols"] = static_cast<uint32_t>(this->cols());
 	out["data"] = this->inMatlabFormat();
 }
 /** Serialize CSchemeArchiveBase derived object to CSerializable Object*/

--- a/libs/math/src/CMatrixD.cpp
+++ b/libs/math/src/CMatrixD.cpp
@@ -24,7 +24,7 @@ uint8_t CMatrixD::serializeGetVersion() const { return 0; }
 void CMatrixD::serializeTo(mrpt::serialization::CArchive& out) const
 {
 	// First, write the number of rows and columns:
-	out << (uint32_t)rows() << (uint32_t)cols();
+	out << static_cast<uint32_t>(rows()) << static_cast<uint32_t>(cols());
 
 	if (rows() > 0 && cols() > 0)
 		for (Index i = 0; i < rows(); i++)
@@ -57,8 +57,8 @@ void CMatrixD::serializeFrom(mrpt::serialization::CArchive& in, uint8_t version)
 void CMatrixD::serializeTo(mrpt::serialization::CSchemeArchiveBase& out) const
 {
 	SCHEMA_SERIALIZE_DATATYPE_VERSION(1);
-	out["nrows"] = this->rows();
-	out["ncols"] = this->cols();
+	out["nrows"] = static_cast<uint32_t>(this->rows());
+	out["ncols"] = static_cast<uint32_t>(this->cols());
 	out["data"] = this->inMatlabFormat();
 }
 /** Serialize CSchemeArchiveBase derived object to CSerializable Object*/

--- a/libs/obs/include/mrpt/obs/CObservationVelodyneScan.h
+++ b/libs/obs/include/mrpt/obs/CObservationVelodyneScan.h
@@ -200,7 +200,7 @@ class CObservationVelodyneScan : public CObservation
 		/** The list of point indices (in x,y,z,...) generated for each laserID
 		 * (ring number). Generated only if `generatePointsForLaserID`=true in
 		 * `TGeneratePointCloudParameters`*/
-		std::vector<std::vector<std::size_t>> pointsForLaserID;
+		std::vector<std::vector<uint64_t>> pointsForLaserID;
 
 		inline std::size_t size() const { return x.size(); }
 		void reserve(std::size_t n);

--- a/libs/opengl/include/mrpt/opengl/COpenGLScene.h
+++ b/libs/opengl/include/mrpt/opengl/COpenGLScene.h
@@ -229,19 +229,6 @@ class COpenGLScene : public mrpt::serialization::CSerializable
 		MRPT_END
 	}
 
-	/** Recursive depth-first visit all objects in all viewports of the scene,
-	 * calling the user-supplied function
-	 *  The passed function must accept a first argument of type "const
-	 * mrpt::opengl::CRenderizable::Ptr &"
-	 *  and a second one of type EXTRA_PARAM
-	 */
-	template <typename FUNCTOR, typename EXTRA_PARAM>
-	inline void visitAllObjects(
-		FUNCTOR functor, const EXTRA_PARAM& userParam) const
-	{
-		visitAllObjects(std::bind2nd(functor, userParam));
-	}
-
    protected:
 	bool m_followCamera{false};
 

--- a/libs/random/include/mrpt/random/RandomGenerators.h
+++ b/libs/random/include/mrpt/random/RandomGenerators.h
@@ -43,8 +43,14 @@ class Generator_MT19937
 {
    public:
 	using result_type = uint32_t;
-	static constexpr result_type min() { return std::numeric_limits<result_type>::min(); }
-	static constexpr result_type max() { return std::numeric_limits<result_type>::max(); }
+	static constexpr result_type min()
+	{
+		return std::numeric_limits<result_type>::min();
+	}
+	static constexpr result_type max()
+	{
+		return std::numeric_limits<result_type>::max();
+	}
 
 	result_type operator()();
 

--- a/libs/random/include/mrpt/random/RandomGenerators.h
+++ b/libs/random/include/mrpt/random/RandomGenerators.h
@@ -43,8 +43,8 @@ class Generator_MT19937
 {
    public:
 	using result_type = uint32_t;
-	static result_type min() { return std::numeric_limits<result_type>::min(); }
-	static result_type max() { return std::numeric_limits<result_type>::max(); }
+	static constexpr result_type min() { return std::numeric_limits<result_type>::min(); }
+	static constexpr result_type max() { return std::numeric_limits<result_type>::max(); }
 
 	result_type operator()();
 

--- a/libs/system/src/os.cpp
+++ b/libs/system/src/os.cpp
@@ -669,7 +669,7 @@ int mrpt::system::executeCommand(
 	stringstream sout;
 	int exit_code = -1;
 
-#ifdef MRPT_OS_LINUX
+#ifndef _WIN32
 	// Run Popen
 	FILE* in;
 	char buff[512];

--- a/libs/system/src/scheduler.cpp
+++ b/libs/system/src/scheduler.cpp
@@ -46,12 +46,7 @@ void mrpt::system::changeCurrentThreadPriority(TThreadPriority priority)
 	// TThreadPriority is defined to agree with numbers expected by Win32 API:
 	SetThreadPriority(GetCurrentThread(), priority);
 #else
-	const pthread_t tid =
-#ifdef MRPT_OS_APPLE
-		reinterpret_cast<long unsigned int>(pthread_self());
-#else
-		pthread_self();
-#endif
+	const pthread_t tid = pthread_self();
 
 	int ret, policy;
 	struct sched_param param

--- a/libs/typemeta/include/mrpt/typemeta/TEnumType.h
+++ b/libs/typemeta/include/mrpt/typemeta/TEnumType.h
@@ -10,6 +10,7 @@
 
 #include <map>
 #include <stdexcept>
+#include <string>
 
 namespace mrpt
 {

--- a/libs/typemeta/include/mrpt/typemeta/TTypeName.h
+++ b/libs/typemeta/include/mrpt/typemeta/TTypeName.h
@@ -11,12 +11,16 @@
 #include <cstdint>
 #include <mrpt/typemeta/static_string.h>
 
+#ifdef __APPLE__
+#include <memory>
+#else
 // frwd decl for TTypeName specialization:
 namespace std
 {
 template <class T>
 class shared_ptr;
 }
+#endif
 
 namespace mrpt
 {

--- a/libs/typemeta/include/mrpt/typemeta/static_string.h
+++ b/libs/typemeta/include/mrpt/typemeta/static_string.h
@@ -15,6 +15,8 @@
 
 #include <mrpt/typemeta/xassert.h>
 
+#include <string>
+
 namespace mrpt
 {
 namespace typemeta

--- a/otherlibs/libcvd/cvd/gl_helpers.h
+++ b/otherlibs/libcvd/cvd/gl_helpers.h
@@ -18,7 +18,7 @@
 #include <windows.h>
 #endif
 
-#ifdef _OSX
+#ifdef __APPLE__
 #include <OpenGL/gl.h>
 #include <OpenGL/glu.h>
 #else

